### PR TITLE
Optimize sfm (reboot)

### PR DESCRIPTION
**Pull request by dPys**
This is a reboot of PR #2488, with fresh commit history + rebased with master (i.e. no more missing utils/multiproc.py file), and a fix on the test per @skoudoro 's request. 

@arokem -- I discovered that when fitting the solver in parallel, the `warm_start` feature was causing numerical divergence. Intuitively, this makes sense because fitting solvers in an embarrassingly parallel manner precludes the possibility of sequentially drawing upon the previous fit estimates for initialization. After disabling, this is no longer an issue, and the results are identical with joblib (when specifying any number of n_procs). And frankly, I propose keeping it disabled in this context anyway because as sklearn notes in their docs: "repeatedly calling an sklearn estimator fit when `warm_start=True` can result in a different solution than when calling fit a single time because of the way the data is shuffled".

Not that it was necessary, but I've also added a `seed` argument to the `fit` method of `SparseFascicleModel` for additional control. 

```
 dipy/reconst/tests/test_sfm.py::test_design_matrix ✓                                                    14% █▌        
 dipy/reconst/tests/test_sfm.py::test_sfm ✓                                                              29% ██▉       
 dipy/reconst/tests/test_sfm.py::test_predict ✓                                                          43% ████▍     
 dipy/reconst/tests/test_sfm.py::test_sfm_background ✓                                                   57% █████▊    
 dipy/reconst/tests/test_sfm.py::test_sfm_stick ✓                                                        71% ███████▎  
 dipy/reconst/tests/test_sfm.py::test_sfm_sklearnlinearsolver ✓                                          86% ████████▋ 
 dipy/reconst/tests/test_sfm.py::test_exponential_iso ✓                                                 100% ██████████
================================================== warnings summary ===================================================
-- Docs: https://docs.pytest.org/en/latest/warnings.html

Results (100.74s):
       7 passed
```

Assuming the disabling of warm_start is acceptable and the CI is happy, then this PR should be ready to merge...

@dPys

### CODE REVIEW

**Comment by skoudoro** [comment]
Thank you for the update @dPys.

Can you close #2488 to avoid any confusion? since you cited it on the description we can easily go back to our previous comments without letting this PR open.
 
I will have a deeper look but overall,  it looks fine.

CI's failures are not related so they can be ignored. Today, I just started the migration to Github Action, so hopefully, I will fix all these errors at the same time.

Thanks and we will get back to you ASAP

**Comment by dPys** [comment]
> Thank you for the update @dPys.
> 
> Can you close #2488 to avoid any confusion? since you cited it on the description we can easily go back to our previous comments without letting this PR open.
> 
> I will have a deeper look but overall, it looks fine.
> 
> CI's failures are not related so they can be ignored. Today, I just started the migration to Github Action, so hopefully, I will fix all these errors at the same time.
> 
> Thanks and we will get back to you ASAP

Cool, done.

File: dipy/reconst/sfm.py [hunk 1]
```diff
@@ -489,22 +580,28 @@ def fit(self, data, mask=None):
...
+
         S0 : ndarray
             The mean non-diffusion-weighted signal.
+
         iso : IsotropicFit class instance
             A representation of the isotropic signal, together with parameters
             of the isotropic signal in each voxel, that is capable of
             deriving/predicting an isotropic signal, based on a gradient-table.
         """
+        super().__init__(self, model)
```
Comment (arokem) [inline dipy/reconst/sfm.py:603]: I believe that if you add this, you can remove the assignment `self.model` below on L605.

File: dipy/reconst/tests/test_sfm.py [hunk 2]
```diff
@@ -2,6 +2,7 @@
 import numpy as np
 import numpy.testing as npt
 import pytest
+import time
```
Comment (arokem) [inline dipy/reconst/tests/test_sfm.py:5]: I don't think this is being used.

**Review by arokem — COMMENTED** [review]
Looks good to me! I had just a couple of really small and easy-to-address comments.

File: dipy/reconst/tests/test_sfm.py (hunk 2 above)
Comment (dPys) [reply dipy/reconst/tests/test_sfm.py:5]: yep, I had included it initially for benchmarking, but will now remove.

File: dipy/reconst/sfm.py (hunk 1 above)
Comment (dPys) [reply dipy/reconst/sfm.py:603]: If I do that, then the following occurs:
```
    def odf(self, sphere):
        """
        The orientation distribution function of the SFM
    
        Parameters
        ----------
        sphere : Sphere
            The points in which the ODF is evaluated
    
        Returns
        -------
        odf :  ndarray of shape (x, y, z, sphere.vertices.shape[0])
    
        """
>       odf_matrix = self.model.cache_get('odf_matrix', key=sphere)
E       AttributeError: 'SparseFascicleFit' object has no attribute 'cache_get'
```
for the subsequent `odf` method.

File: dipy/reconst/tests/test_sfm.py (hunk 2 above)
Comment (dPys) [reply dipy/reconst/tests/test_sfm.py:5]: yep, removed.

File: dipy/reconst/sfm.py (hunk 1 above)
Comment (arokem) [reply dipy/reconst/sfm.py:603]: OK - I'm probably wrong then!

**Comment by arokem** [comment]
Assuming the CI failures are unrelated, and no other errors appear, I believe this is ready to go!

**Comment by skoudoro** [comment]
Thanks @dPys, merging !

**Pull request merged**